db/migrate/201808131137612_policy_log_trigger_bypass.rb

```diff
@@ -2,13 +2,48 @@
 
 Sequel.migration do
   tables = %i(roles role_memberships resources permissions annotations)
-
+  
   up do
+    # `public_log_record` is a custom type to record the column
+    # definitions for the policy log records generated by each of
+    # the `policy_log_#{table}_record` functions
+    execute <<-SQL
+    CREATE TYPE policy_log_record as (
+      policy_id text,
+      version int,
+      operation policy_log_op,
+      kind policy_log_kind,
+      subject hstore);
+    SQL
 
     tables.each do |table|
       # find the primary key of the table
       primary_key = schema(table).select{|x,s|s[:primary_key]}.map(&:first).map(&:to_s).pg_array
-      execute """
+
+      # Each table has a function format the policy record for that type.
+      # This function is used both by the policy_log trigger for each
+      # table, as well as for the bulk insert policy log in the policy
+      # load orchestrator.
+      execute <<-SQL
+        CREATE OR REPLACE FUNCTION policy_log_#{table}_record(
```

Comment:
Maybe, instead of creating a function for each table, we can have a single function that takes the table name as an argument? This SO answer seems like it describes a reasonable technique: https://stackoverflow.com/a/10711349 . (I did read the objection in the subsequent answer, and I'm ok with disregarding the advice in it.)

I don't really understand the other problem you asked about, but maybe this approach will get around it. It not, can you describe it in more detail (e.g. are the errors you get when don't marshal through `text`)?